src/texture_cache.rs

```diff
@@ -799,100 +799,86 @@ impl TextureCache {
-        let location = page_list.last_mut().and_then(|last_page| {
-            last_page.allocate(&allocation_size, filter)
-        });
-        let location = match location {
-            Some(location) => location,
-            None => {
-                if !page_list.is_empty() && page_list.last().unwrap().can_grow() {
-                    let last_page = page_list.last_mut().unwrap();
-                    // Grow the texture.
-                    let texture_size = cmp::min(last_page.texture_size * 2,
-                                                max_texture_size());
-                    self.pending_updates.push(TextureUpdate {
-                        id: last_page.texture_id,
-                        op: texture_grow_op(texture_size, format, mode),
-                    });
-                    last_page.grow(texture_size);
-
-                    self.items.for_each_item(|item| {
-                        if item.texture_id == last_page.texture_id {
-                            item.texture_size = Size2D::new(texture_size, texture_size);
-                        }
-                    });
-                } else {
-                    // We need a new page.
-                    let texture_size = initial_texture_size();
-                    let free_texture_levels_entry = match kind {
-                        TextureCacheItemKind::Standard => self.free_texture_levels.entry(format),
-                        TextureCacheItemKind::Alternate => {
-                            self.alternate_free_texture_levels.entry(format)
-                        }
-                        //TextureCacheItemKind::RenderTarget => {
-                        //    self.render_target_free_texture_levels.entry(format)
-                        //}
-                    };
-                    let mut free_texture_levels = match free_texture_levels_entry {
-                        Entry::Vacant(entry) => entry.insert(Vec::new()),
-                        Entry::Occupied(entry) => entry.into_mut(),
-                    };
-                    if free_texture_levels.is_empty() {
-                        create_new_texture_page(&mut self.pending_updates,
-                                                &mut self.free_texture_ids,
-                                                &mut free_texture_levels,
-                                                texture_size,
-                                                format,
-                                                mode);
+
+        // TODO(gw): Handle this sensibly (support failing to render items that can't fit?)
+        assert!(allocation_size.width < max_texture_size());
+        assert!(allocation_size.height < max_texture_size());
+
+        loop {
+            let location = page_list.last_mut().and_then(|last_page| {
+                last_page.allocate(&allocation_size, filter)
+            });
+
+            match location {
+                Some(location) => {
```

Comment:
nit: How about `if let Some(location) = location { ... }`? Since there's a `return` at the end, this should be equivalent, and then you can eliminate lots of rightward drift.